cvmfs/swissknife_filestats.cc

```diff
@@ -0,0 +1,317 @@
+}
+
+template <class ObjectFetcherT>
+bool CommandFileStats::Run(ObjectFetcherT *object_fetcher)
+{
+  running_ = true;
+
+  string abs_path = GetAbsolutePath(db_path_);
+  unlink(abs_path.c_str());
+  db_ = FileStatsDatabase::Create(db_path_);
+  db_->InitStatements();
+
+  RemoveTree(tmp_db_path_);
+  assert(MkdirDeep(tmp_db_path_, 0755));
+
+  typename CatalogTraversal<ObjectFetcherT>::Parameters params;
+  params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(params);
+  traversal.RegisterListener(&CommandFileStats::CatalogCallback, this);
+
+  pthread_create(&thread_processing_, NULL, MainProcessing, this);
+
+  bool ret = traversal.Traverse();
+
+  running_ = false;
+  pthread_join(thread_processing_, NULL);
+
+  db_->DestroyStatements();
+
+  return ret;
+}
+
+void CommandFileStats::CatalogCallback(
+  const CatalogTraversalData<catalog::Catalog> &data) {
+  string out_path =  tmp_db_path_ + StringifyInt(num_downloaded_ + 1) + ".db";
+  assert(CopyPath2Path(data.catalog->database_path(), out_path));
+  ++num_downloaded_;
+}
+
+void *CommandFileStats::MainProcessing(void *data) {
+  CommandFileStats *repo_stats = static_cast<CommandFileStats *>(data);
+  int processed = 0;
+  while (repo_stats->running_ || processed < repo_stats->num_downloaded_) {
+    if (processed < repo_stats->num_downloaded_) {
+      LogCvmfs(kLogCatalog, kLogStdout, "Processing catalog %d", processed);
+      string db_path = repo_stats->tmp_db_path_ + "/" +
+                       StringifyInt(processed + 1) + ".db";
+      repo_stats->ProcessCatalog(db_path);
+      unlink(db_path.c_str());
+      ++processed;
+    }
+  }
+  return NULL;
+}
+
+
+
+void CommandFileStats::ProcessCatalog(string db_path) {
+  sqlite::Database<catalog::CatalogDatabase> *cat_db;
+  cat_db = sqlite::Database<catalog::CatalogDatabase>::Open(
+           db_path,
+           sqlite::Database<catalog::CatalogDatabase>::kOpenReadOnly);
+  int64_t file_size = GetFileSize(db_path);
+  sqlite::Sql *catalog_count = new sqlite::Sql(cat_db->sqlite_db(),
+                                               "SELECT count(*) FROM catalog;");
+  catalog_count->Execute();
+  int cur_catalog_id = db_->StoreCatalog(catalog_count->RetrieveInt64(0),
+                                         file_size);
+  delete catalog_count;
+
+  sqlite::Sql *catalog_list =
+    new sqlite::Sql(cat_db->sqlite_db(),
+                    "SELECT hash, size, flags, symlink FROM catalog;");
+  sqlite::Sql *chunks_list =
+    new sqlite::Sql(cat_db->sqlite_db(),
+                    "SELECT md5path_1, md5path_2, size, hash FROM chunks "
+                    "ORDER BY md5path_1 ASC, md5path_2 ASC;");
+
+  db_->BeginTransaction();
+
+  while (catalog_list->FetchRow()) {
+    const void *hash = catalog_list->RetrieveBlob(0);
+    int num_bytes = catalog_list->RetrieveBytes(0);
+    int64_t size = catalog_list->RetrieveInt64(1);
+    int flags = catalog_list->RetrieveInt(2);
+    if (flags == 4) {
```

Comment:
Use symbolic flag values

Reply:
Flags are combinable, so most likely this should query `(flags & VALUE) == VALUE`